### PR TITLE
C++: Add more documentation about dataflow through indirections

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR adds a new article (linked from https://codeql.github.com/docs/codeql-language-guides/codeql-for-cpp/) explaining how we would tackle a few situations that involve tainting a pointer or tainting its indirection.

We know from past experience that this has been something Microsoft has needed some explanation of, and this document should hopefully answer some of the common questions we've gotten (with the expectation that other people will have similar questions).

cc @felicitymay

### CODE REVIEW

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+
+Advanced dataflow scenarios for C/C++
+======================================
+
+Data flow for C and C++ distinguish between the value of a pointer and the value of what the pointer points to. We call this the "indirection" of the pointer. Tracking the pointer and its indirection as separate entities is important for precise dataflow, but it may mean that you do not get the right flow if you select the wrong data flow node. This article demonstrates various scenarios where it is important to consider whether data flow should be computed on the value of the pointer or its indirection.
+
+For almost all situations we only need to instantiate a dataflow configuration and specify our sources and sinks, and dataflow will handle everything for us.
+
+However, when a write to a field is not visible to CodeQL (for example because it happens in a function whose definition is not in the database) we need to track the qualifier, and tell the dataflow library that it is okay to transfer from flow from the qualifier to the field access. This is done by adding an ``isAdditionalFlowStep`` predicate to the dataflow module.
```
Comment (geoffw0) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:14]:
```suggestion
However, when a write to a field is not visible to CodeQL (for example because it happens in a function whose definition is not in the database) we need to track the qualifier, and tell the dataflow library that it is okay to transfer flow from the qualifier to the field access. This is done by adding an ``isAdditionalFlowStep`` predicate to the dataflow module.
```

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+will catch most things á la
```
Comment (geoffw0) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:72]: > á la

We should probably rephrase this to something more "basic English".

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+
+  struct B {
+    A *a;
+    int y;
+  };
+
+  void sink(int);
+  int user_input();
+  void write_user_input_to(void *);
+  void read_data(const void *);
```
Comment (geoffw0) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:40]: This is a lot to take in, and we don't actually use all of it right away.  Can we consider moving the definitions of `A` and `B` to the block where they are first used?

Also `write_user_input_to` and `read_data` are, I think, introduced inline before they are used so are perhaps redundant here?

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+  }
+
+  void get_and_process_user_input_v2() {
+    U* u = (U*)malloc(sizeof(U));
+    write_user_input_to(u);
+    process_user_data(u->p);
+    free(u);
+  }
+
+Here, flow starts at the outgoing argument of ``write_user_input_to(...)`` and proceeds to ``u->p``. However, because CodeQL has not observed a write to p prior to the read ``u->p``, dataflow will stop at ``u``. In order to convince CodeQL to proceed we need to add an additional flow step through field reads like so:
```
Comment (geoffw0) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:129]: > Here, flow starts at the outgoing argument of ``write_user_input_to(...)``

I doesn't though, because you haven't introduced `write_user_input_to` into `isSource` yet.  I think we should consider moving the QL below that adds the source and sink (but perhaps not the additional flow step) to just above this paragraph?

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+  int user_input();
+  void read_data(const void *);
+  void *malloc(size_t);
+
+  void get_input_and_use_data() {
+    B b;
+    b.a = (A *)malloc(sizeof(A));
+    b.a->x = user_input();
+    // ...
+    use_data(&b);
```
Comment (geoffw0) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:370]:
```suggestion
    read_data(&b);
```
I think `use_data` and `read_data` are references to the same thing?

**Review by geoffw0 — COMMENTED** [review]
Just read this through.  Looks fantastic and very valuable to advanced users!  I do have a fair number of minor corrections and some suggestions that I think may improve readability.  Feel free to accept / reject / change / discuss my suggestions as you see fit.

I have not checked all the links work.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 2 above)
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Good idea. I've replaced it with `such as` in 3de32e813b5e5fa46fc9990a86b4dae8503f472d. Is that better?

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 3 above)
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: I agree. In 3de32e813b5e5fa46fc9990a86b4dae8503f472d I've reworded it so that it only introduces `user_input` and `sink`. Does that improve the situation?

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+    // ...
+    process_structure(b);
+    free(b);
+  }
+
+This is simple to match because we see:
+  1. User input starts at ``user_input()`` and flows into ``fill_structure``.
+  2. The data is written to the object ``b`` with access path ``[a, p]``.
+  3. The object ``b`` flows out of ``fill_structure`` and into ``process_structure``.
+  4. The access path ``[a, p]`` is read in ``process_structure`` and the value ends up in the sink.
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:96]: We could definitely do that. I wanted to convey the fact that the "standard thing" doesn't just work for simple things where we store an integer in a field and later read it out, but also handles the fact that we're storing a pointer to something and later dereference it (because that was a misconception we had to explain to Microsoft in the past).

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 4 above)
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: That's a good point. I've tried to reorganize things in d6ccd63ed2db16b4c47f298bf00d0bcb64c9142c so that the relevant `isSource` is mentioned first. Does that improve the readability?

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 5 above)
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Oops. Indeed they are. I had to fix a few other references, so I did it in c5307bed2b3c691c793b19218afe32b26b96d191 instead of accepting your commit suggestion.

**Comment by felicitymay** [comment]
I'm going to hold off on starting the writer review until you let me know that the technical review is complete.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+        call.getTarget().hasName("use_value") and
+        sink.asExpr() = call.getAnArgument()
+      )
+    }
+
+    predicate isAdditionalFlowStep(DataFlow::Node n1, DataFlow::Node n2) {
+      exists(FieldAccess fa |
+        n1.asIndirectExpr() = fa.getQualifier() and
+        n2.asIndirectExpr() = fa
+      )
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:167]: I've added a small comment about this in 0a71705dfe5f2fc73d37e1f05a2fe9ae99504ccd. Let me know if you think we should actually do the optimization you suggested instead of simply mentioning it in the prose.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+    b.a = (A *)malloc(sizeof(A));
+    b.a->x = user_input();
+    // ...
+    use_data(&b);
+    free(b.a);
+  }
+
+We write a user-controlled value into the object ``b`` at the access path ``[a, x]``. Afterwards, ``b`` is passed to ``use_data`` which we assume we don't have the definition of in the database. We now want to track this user-input flowing into ``use_data``.
+
+The dataflow library actually has a specific tool to handle this scenario, and thus we don't need to add any additional flow steps using ``isAdditionalFlowStep`` to handle this. Instead, we have to tell the dataflow library that ``use_data`` may implicitly read the data from the object that it has been passed. To do that, we implement ``allowImplicitRead`` in our dataflow module:
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Good idea. I've reformulated this in c4075b3ec7d2889e02392c5be401edb20fa0ac21.

(The `read_data` / `use_data` was fixed in c5307bed2b3c691c793b19218afe32b26b96d191.)

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+The ``allowImplicitRead`` predicate specifies that if we're at a node that satisfies ``isSink`` then we're allowed to assume that there is an implicit read of a field named ``a`` or a field named ``x``. This gets us the flow we are interested in because the dataflow library now will see:
+
+  1. User input starts at ``user_input()``.
+  2. The data flowing into ``b`` with access path ``[a, x]``.
+  3. The data flowing to the indirection of ``&b`` (i.e., the object ``b``).
+  4. An implicit read of the field ``x`` followed by an implicit read of the field ``a``.
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:431]: Yep, I like that change 👍

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+The ``allowImplicitRead`` predicate specifies that if we're at a node that satisfies ``isSink`` then we're allowed to assume that there is an implicit read of a field named ``a`` or a field named ``x``. This gets us the flow we are interested in because the dataflow library now will see:
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:426]: Yeah, we can definitely do that 👍

**Comment by MathiasVP** [comment]
@geoffw0 thanks for all the great feedback! I think I've incorporated all the changes that I wanted to incorporate 😅. I think the only one I left out was https://github.com/github/codeql/pull/14524#discussion_r1362087824 because I do think it serves a purpose to show that "standard dataflow" works well even for complex cases such as this.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 4 above)
Comment (geoffw0) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Yes, much better!

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:168]: The added comment is definitely good.  I would be tempted to implement it in the code as well (because people love to copy-paste code from tutorials), but I can see that there's a design decision involved in whether you're matching a type name or field or whatever so it isn't one size fits all.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 3 above)
Comment (geoffw0) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Yep, I'm much happier with that.

I think when I read a code block I expect to be able to understand what is going on in it without reading ahead [far] - that's why the introduction of `A` and `B` "too early" were bothering me.

:+1:

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:168]: I don't have strong opinions about this though.  It's fine as it is (now).

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst (hunk 6 above)
[the remark this answers is not in the document]
Comment (geoffw0) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:97]: I see your point.  Happy with this as it is.

**Comment by felicitymay** [comment]
I'll take a look tomorrow morning 👍🏻

**Review by felicitymay — COMMENTED** [review]
Many thanks for taking the time to write this detailed article. It looks as if the examples of different ways to track pointers and indirections will be really useful in helping query writers understand their options. 💖 

I've added a lot of small text suggestions and a few other thoughts on how this could be clearer. As always, please take care to ensure that the suggested changes don't change the meaning, and remember that I'm not in the target audience for this article so feel free to reject suggestions 🙂

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+
+For almost all situations we only need to instantiate a dataflow configuration and specify our sources and sinks, and dataflow will handle everything for us.
+
+However, when a write to a field is not visible to CodeQL (for example because it happens in a function whose definition is not in the database) we need to track the qualifier, and tell the dataflow library that it is okay to transfer flow from the qualifier to the field access. This is done by adding an ``isAdditionalFlowStep`` predicate to the dataflow module.
+
+When adding additional flow steps involving pointers one needs to be careful about whether the dataflow step should flow from the pointer or its indirection. Similarly, care must be taken to decide whether the additional step should target a pointer or its indirection.
+
+In the dual situation where a read of a field is not visible to CodeQL, the dataflow library's "implicit read" mechanism can be used to achieve the right flow.
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:21]:
```suggestion
In contrast, if the read of a field is not visible to CodeQL, you can add an ``allowImplicitRead`` predicate to model the data flow.
```

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+    }
+  }
+
+  module Flow = DataFlow::Global<Config>;
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+will catch most things such as
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:63]: This is meant to refer to the subsequent code block. So I think you meant to suggest a colon instead?

```suggestion
will catch most things such as:
```

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+  }
+
+  void get_and_process_user_input_v2() {
+    U* u = (U*)malloc(sizeof(U));
+    write_user_input_to(u);
+    process_user_data(u->p);
+    free(u);
+  }
+
+Flow now starts at the outgoing argument of ``write_user_input_to(...)`` and proceeds to ``u->p``. However, because CodeQL has not observed a write to ``p`` prior to the read ``u->p``, dataflow will stop at ``u``. In order to convince CodeQL to proceed we need to add an additional flow step through field reads like so:
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: I've converted this one to a numbered list in 1b1a78bf044f6a7b62c79978af343de1127b8c0d. I'll go through the other ones and see if there are similar opportunities.

I also incorporated your suggested changes to the unchanged parts of the paragraph. Thank you!

[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: f6b1d66167289f961bca4097397dba785c6fdddb was another meaningful change I could find like the above. I don't think it makes sense to convert any of the other paragraphs to numbered lists, though. Thanks for the suggestion! 🙇

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+
+  from Flow::PathNode source, Flow::PathNode sink
+  where Flow::flowPath(source, sink)
+  select sink.getNode(), source, sink, "Flow from user input to sink!"
+
+Notice how the ``isSource`` and ``isSink`` are as expected: we're looking for flow that starts at the outgoing parameter of ``write_user_input_to(...)``, and ends up as an argument to ``isSink``. The interesting part is the addition of ``isAdditionalFlow`` which specifies an additional flow step from the qualifier of a ``FieldAccess`` to the result of the access.
+
+In a real query the ``isAdditionalFlowStep`` step would be restricted in various ways to make sure that it doesn't add too much flow (since flow from a field qualifier to the field access in general will generate a lot of spurious flow). For example, one could restrict ``fa`` to be a field access that targets a particular field, or a field access of a field that's defined in a certain ``struct`` type.
+
+We have an important choice here: Should the relationship between ``n2`` and ``fa`` be implemented using ``asExpr`` or ``asIndirectExpr``?
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Good idea. I've fixed this in 44e214b9337f300b1ba802f7279aa9af2abb2546.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+      sink.asIndirectExpr() = call.getAnArgument()
+    )
+  }
+
+.. _using-asExpr:
+
+Using asExpr
+~~~~~~~~~~~~
+
+Alternatively, this flow could also be tracked by:
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: I've added the named code blocks as you suggested [here](https://github.com/github/codeql/pull/14524#discussion_r1363727847) and hopefully used that to make it more clear in c04546da98052d0a9b15bfba4233b4b90eeeb89d which code snippet it is referring to 🤞.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+We have an important choice here: Should the relationship between ``n2`` and ``fa`` be implemented using ``asExpr`` or ``asIndirectExpr``? 
+
+.. _using-asIndirectExpr:
+
+Using asIndirectExpr
+~~~~~~~~~~~~~~~~~~~~
+
+If we use ``n2.asIndirectExpr() = fa`` we specify that flow moves to what ``fa`` points to. This allows dataflow to flow through a later dereference, which is exactly what we need to to flow from ``p`` to ``*p`` in ``process_user_data``.
+
+Thus we get the required flow path.
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:201]: Yes, it's referring to the previous example. I've hopefully made this more explicit in c04546da98052d0a9b15bfba4233b4b90eeeb89d with the use of captions 🤞.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+
+
+  void get_and_process_user_input_v2() {
+    U* u = (U*)malloc(sizeof(U));
+    write_user_input_to(u);
+    process_user_data(u->data);
+    free(u);
+  }
+
+Since data is no longer a pointer our ``isAdditionalFlowStep`` doesn't make any sense because it specifies flow to the indirection of the field (and an integer does not have any indirections). So there is no choice about whether to taint the value of the field or its indirection: it has to be the value. However, since we pass the address of ``data`` to ``use_pointer`` the tainted data is what is pointed to by the argument of ``use_pointer`` (since the data pointed to by ``&data`` is exactly ``data``). So to handle this case we need a mix of the two situations above:
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst]: Indeed this is technically a complex paragraph. I've tried to clear up the explanation in a36d432184cf015f81ea2f3cb606f4e2530e9c18.

**Comment by MathiasVP** [comment]
@felicitymay thank you for all the suggestions. They certainly made the article a lot nicer to read! I believe I've fixed all your comments now 🤞

**Comment by felicitymay** [comment]
> I believe I've fixed all your comments now 🤞

Many thanks for working through all the comments 💖 

Sadly I have meetings the rest of the day, but I will take a last look tomorrow morning.

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 18]
```diff
@@ -179,20 +191,22 @@ Notice how the ``isSource`` and ``isSink`` are as expected: we're looking for fl
 
 In a real query the ``isAdditionalFlowStep`` step would be restricted in various ways to make sure that it doesn't add too much flow (since flow from a field qualifier to the field access in general will generate a lot of spurious flow). For example, one could restrict ``fa`` to be a field access that targets a particular field, or a field access of a field that's defined in a certain ``struct`` type.
 
-We have an important choice here: Should the relationship between ``n2`` and ``fa`` be implemented using ``asExpr`` or ``asIndirectExpr``? 
+We have an important choice here: Should ``n2`` be the node corresponding to the pointer value of ``fa`` or the indirection of ``fa`` (i.e., what ``fa`` points to)?
```
Comment (felicitymay) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:194]: We avoid using this shorthand.

```suggestion
We have an important choice here: Should ``n2`` be the node corresponding to the pointer value of ``fa`` or the indirection of ``fa`` (that is, what ``fa`` points to)?
```

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 19]
```diff
@@ -306,9 +322,11 @@ Consider an alternative scenario where ``U`` contains a single ``int`` data, and
...
-  1. We need to taint the value of the field just like in the :ref:`Using asExpr <using-asExpr>` section.
-  2. We need to select the indirection of the argument just like in the :ref:`Using asIndirectExpr <using-asIndirectExpr>` section.
+Since the ``data`` field is now an ``int`` instead of an ``int*`` the field no longer has any indirections, and so the use of ``asIndirectExpr`` in ``isAdditionalFlowStep`` no longer makes sense (and so the additional step will have no results). So there is no choice about whether to taint the value of the field or its indirection: it has to be the value.
+
+However, since we pass the address of ``data`` to ``use_pointer`` on line 12 the tainted value is what is pointed to by the argument of ``use_pointer`` (since the value pointed to by ``&data`` is exactly ``data``). So to handle this case we need a mix of the two situations above:
+  1. We need to taint the value of the field as described the :ref:`Using asExpr <using-asExpr>` section.
```
Comment (felicitymay) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:328]: Missing word - I think this was missing in my last suggestion for this line :disappointed:

```suggestion
  1. We need to taint the value of the field as described in the :ref:`Using asExpr <using-asExpr>` section.
```

File: docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst [hunk 20]
```diff
@@ -384,9 +404,14 @@ To set the stage, consider the following scenario:
     free(b.a);
   }
 
-We write a user-controlled value into the object ``b`` at the access path ``[a, x]``. Afterwards, ``b`` is passed to ``read_data`` which we don't have the definition of in the database. We now want to track this user-input flowing into ``read_data``.
+What happens here is the following:
```
Comment (felicitymay) [inline docs/codeql/codeql-language-guides/advanced-dataflow-scenarios-cpp.rst:407]: Suggest:

```suggestion
In this example, the data flows as follows:
```

**Review by felicitymay — APPROVED** [review]
Many thanks for all the updates. I found a couple of very small nits, but otherwise this looks ready to merge.

I particularly like the way you've formatted the C/C++ examples.

**Review by felicitymay — APPROVED** [review]

**Comment by felicitymay** [comment]
@MathiasVP - we're just about to start the update process for the CodeQL docs site for the release of the CLI. If you'd like this to be published sooner, you could retarget this against the release candidate branch. Otherwise, this is likely to be published in a couple of weeks.
